### PR TITLE
Add typecheck and fix errors

### PR DESCRIPTION
**Pull request by miurahr**
Re-create commit history based on #617  for merge.

### CODE REVIEW

File: aqt/metadata.py [hunk 1]
```diff
@@ -591,16 +595,14 @@ def fetch_arches(self, version: Version) -> List[str]:
...
-        def get_version(ver_ext: Tuple[Version, str]):
-            return ver_ext[0]
+            return version is not None and (self.spec is None or version in self.spec) and (ext == extension)
 
         versions_extensions = self.get_versions_extensions(
             self.fetch_http(self.archive_id.to_url(), False), self.archive_id.category
         )
-        versions = sorted(filter(None, map(get_version, filter(filter_by, versions_extensions))))
-        iterables = itertools.groupby(versions, lambda version: version.minor)
+        opt_versions = map(lambda _tuple: _tuple[0], filter(filter_by, versions_extensions))
```
Comment (miurahr) [inline aqt/metadata.py:603]: snyk analyzer warns "Combine filter then map into one comprehension" can be "[x for x in y if z]" style.
```suggestion
        opt_versions = [_tuple[0] for _tuple in filter(filter_by, versions_extensions)]
```

File: aqt/archives.py [hunk 2]
```diff
@@ -23,6 +23,7 @@
 from dataclasses import dataclass, field
 from logging import getLogger
 from typing import Dict, Iterable, List, Optional, Tuple
+from xml.etree.ElementTree import Element
```
[the remark this answers is not in the document]
Comment (miurahr) [reply aqt/archives.py:26]: @sonatype-lift ignore

File: aqt/metadata.py [hunk 3]
```diff
@@ -591,16 +595,14 @@ def fetch_arches(self, version: Version) -> List[str]:
...
+        versions: List[Version] = sorted(filter(None, opt_versions))
```
Comment (miurahr) [inline aqt/metadata.py:604]:
```suggestion
        versions: List[Version] = sorted([x for x in filter(None, opt_versions)])
```

File: aqt/metadata.py [hunk 4]
```diff
@@ -635,7 +637,7 @@ def choose_highest_version_in_spec(
             return None
 
         # Remove items that don't conform to simple_spec
-        tools_versions = filter(lambda tool_item: tool_item[2] in simple_spec, tools_versions)
+        tools_versions = list(filter(lambda tool_item: tool_item[2] in simple_spec, tools_versions))
```
Comment (miurahr) [inline aqt/metadata.py:640]:
```suggestion
        tools_versions = [tool_item for tool_item in tools_versions if tool_item[2] in simple_spec]
```

**Comment by ddalcino** [comment]
This commit will fix the last mypy error: https://github.com/miurahr/aqtinstall/pull/617/commits/54c3c4a3e89a3246f753dd25851439d4c14bf414

**Comment by miurahr** [comment]
I'd like to merge these 11 commits.
Further improvements will be done on another issues/PRs.

**Pull request merged**